### ISSUE TITLE
Performance regression for reduction on GPU

### ISSUE THREAD

**Issue by rabbott999** [google/jax#18424]
### Description

I've observed a performance regression moving from jax 0.4.14 to 0.4.16 in a rejection sampling while loop. Here's a simplified version of the code that exhibits the same behavior:

```python
import timeit

import jax

import jax.numpy as jnp


def rejection_sample(sample_fn, accept_fn, *, rng_key):
    def loop_body(state):
        (rng_key, sample, accepted_mask) = state
        rng_key, subkey = jax.random.split(rng_key)
        new_sample = sample_fn(subkey)
        sample = jax.tree_map(
            lambda x, x_new: jnp.where(accepted_mask, x, x_new),
            sample, new_sample)
        accepted_mask = jnp.logical_or(accept_fn(new_sample),
                                       accepted_mask)
        return rng_key, sample, accepted_mask

    def loop_cond(state):
        _, _, accepted_mask = state
        return jnp.logical_not(jnp.all(accepted_mask))

    rng_key, subkey = jax.random.split(rng_key)
    first_sample = sample_fn(subkey)
    first_accepted = accept_fn(first_sample)
    _, sample, _ = jax.lax.while_loop(loop_cond, loop_body,
                                      (rng_key, first_sample,
                                       first_accepted))
    return sample


def sample(shape, *, rng_key):
    def sample_fn(rng_key):
        return jax.random.uniform(rng_key, minval=0.0, maxval=1.0,
                                  shape=shape)

    def accept_fn(sample):
        return sample < 0.7

    return rejection_sample(sample_fn, accept_fn, rng_key=rng_key)


L = 16
Lt = 48
rng_key = jax.random.PRNGKey(0)
shape = (L, L, L, Lt)

# Force compilation before timing
sample_jit = jax.jit(lambda key: sample(shape, rng_key=key))
rng_key, subkey = jax.random.split(rng_key)
x0 = sample_jit(subkey).block_until_ready()

rng_key, subkey = jax.random.split(rng_key)
dt = timeit.timeit(lambda: sample_jit(subkey).block_until_ready(),
                   number=100)
print(dt)
```

Running this under several different environments yields

```
Using jax-0.4.10
0.07284442099989974
Using jax-0.4.12
0.08756969699970796
Using jax-0.4.14
0.10284504599985667
Using jax-0.4.16
7.349518729002739
Using jax-0.4.20
8.398614255998837
```

I.e. the newer version is ~100x slower (in fact, there seems to be a consistent increase in runtime from version to version). I see similar behavior on two different types NVIDIA GPUs (A100 and RTX 2080 Ti). If I force everything to run on CPU, the timings become more inconsistent, but generally consistent with each other, so this likely a GPU-specific issue:

```
Using jax-0.4.10 (cpu)
5.222352617987781
Using jax-0.4.12 (cpu)
4.833697886992013
Using jax-0.4.14 (cpu)
3.9580981749895727
Using jax-0.4.16 (cpu)
5.534542325010989
Using jax-0.4.20 (cpu)
4.845710285997484
```

Potentially related issues:
https://github.com/google/jax/issues/16661
https://github.com/google/jax/issues/16663

### What jax/jaxlib version are you using?

Several (see above)

### Which accelerator(s) are you using?

CPU, GPU

### Additional system info

Python 3.10, Linux

**Comment by patrick-kidger** [comment]
Can you try updating to the latest version of JAX? I believe jaxlib==0.4.16 has a known performance issue that may be related: https://github.com/google/jax/pull/17724

**Comment by rabbott999** [comment]
The latest version I've tried is 0.4.20, which is the most recent release as far as I can tell. The issue is still present there.

The issue you linked inspired me to dig a bit deeper, and after some further investigation it appears the issue is not actually in the while loop, but rather in the condition. Specifically the combination `jnp.logical_not(jnp.all(...))` seems to be the slow part. This simplifies things a bit:

```python
import timeit
import jax
import jax.numpy as jnp

@jax.jit
def f(x):
    return jnp.logical_not(jnp.all(x < 0.7))

@jax.jit
def g(x):
    return jnp.any(jnp.logical_not(x < 0.7))

@jax.jit
def h(x):
    return jnp.all(x < 0.7)


shape = (16**3 * 48,)
rng_key = jax.random.PRNGKey(0)
x = jax.random.uniform(rng_key, minval=0.0, maxval=1.0,
                       shape=shape)

# Compile
f(x)
g(x)
h(x)

dt = timeit.timeit(lambda: f(x).block_until_ready(),
                   number=1000)
print(f"not(all(x)): {dt}")
dt = timeit.timeit(lambda: g(x).block_until_ready(),
                   number=1000)
print(f"any(not(x)): {dt}")
dt = timeit.timeit(lambda: h(x).block_until_ready(),
                   number=1000)
print(f"all(x): {dt}")
```

Running this I get

```
Using jax-0.4.10
not(all(x)): 0.04904459699173458
any(not(x)): 0.040804433010634966
all(x): 0.03323979300330393
Using jax-0.4.12
not(all(x)): 0.06948921899311244
any(not(x)): 0.041380729991942644
all(x): 0.03729355499672238
Using jax-0.4.14
not(all(x)): 0.07314369500090834
any(not(x)): 0.065828587001306
all(x): 0.06242232900694944
Using jax-0.4.16
not(all(x)): 5.161174478009343
any(not(x)): 0.1380664679891197
all(x): 0.11088073400605936
Using jax-0.4.20
not(all(x)): 3.935028567997506
any(not(x)): 0.11883646699425299
all(x): 0.09598468799958937
```

I'm not sure if this is the same as #17724, but it does sound plausible.

**Comment by hawkinsp** [comment]
I can't reproduce this on my (CPU) machine:

```
not(all(x)): 0.031113745993934572
any(not(x)): 0.029049667005892843
all(x): 0.029495747003238648
```

On what platform does this issue occur?

**Comment by rabbott999** [comment]
The issue is only present on GPU -- in particular I've seen this with 2 different Nvidia GPUs (A100 and RTX 2080 Ti), and also two different CUDA versions (11.8 and 12.0).

**Comment by hawkinsp** [comment]
Thanks, it looks like XLA's code generation regressed for this reduction. Filed https://github.com/openxla/xla/issues/7152

**Comment by akuegel** [comment]
We have found that the regression came from reduction epilogue fusion which was implemented in that time frame when the regression occurred. We allowed to create a fusion that is actually not supported by the reduction emitter, so the reduction got emitted via the loop emitter as elementwise reduction, which is quite a bit slower in this case. I am about to submit a fix where we do not allow this fusion, so the reduce will still be the root of the fusion.
